src/lib/logical_query_plan/lqp_utils.cpp

```diff
@@ -0,0 +1,126 @@
+}
+
+std::optional<LQPMismatch>
+lqp_find_structure_mismatch(const std::shared_ptr<const AbstractLQPNode>& lhs, const std::shared_ptr<const AbstractLQPNode>& rhs) {
+  if (!lhs && !rhs) return std::nullopt;
+  if (!(lhs && rhs) || lhs->type != rhs->type) return LQPMismatch(lhs, rhs);
+
+  const auto mismatch_left = lqp_find_structure_mismatch(lhs->left_input(), rhs->left_input());
+  if (mismatch_left) return mismatch_left;
+
+  return lqp_find_structure_mismatch(lhs->right_input(), rhs->right_input());
+}
+
+std::optional<LQPMismatch>
+lqp_find_subplan_mismatch_impl(const LQPNodeMapping& node_mapping, const std::shared_ptr<const AbstractLQPNode>& lhs, const std::shared_ptr<const AbstractLQPNode>& rhs) {
+  if (!lhs && !rhs) return std::nullopt;
+  if (!lhs->shallow_equals(*rhs, node_mapping)) return LQPMismatch(lhs, rhs);
+
+  const auto mismatch_left = lqp_find_subplan_mismatch_impl(node_mapping, lhs->left_input(), rhs->left_input());
+  if (mismatch_left) return mismatch_left;
+
+  return lqp_find_subplan_mismatch_impl(node_mapping, lhs->right_input(), rhs->right_input());
+}
+
+}  // namespace
+
+namespace opossum {
+
+LQPNodeMapping lqp_create_node_mapping(
+const std::shared_ptr<AbstractLQPNode>& lhs, const std::shared_ptr<AbstractLQPNode>& rhs
+) {
+  LQPNodeMapping mapping;
+  lqp_create_node_mapping_impl(mapping, lhs, rhs);
+  return mapping;
+}
+
+std::optional<LQPMismatch>
+lqp_find_subplan_mismatch(const std::shared_ptr<AbstractLQPNode>& lhs, const std::shared_ptr<AbstractLQPNode>& rhs) {
+  // Check for type/structural mismatched
+  auto mismatch = lqp_find_structure_mismatch(lhs, rhs);
+  if (mismatch) return mismatch;
+
+  const auto node_mapping = lqp_create_node_mapping(lhs, rhs);
+
+  return lqp_find_subplan_mismatch_impl(node_mapping, lhs, rhs);
+}
+
+void lqp_replace_node(const std::shared_ptr<AbstractLQPNode>& original_node,
+                      const std::shared_ptr<AbstractLQPNode>& replacement_node) {
+  DebugAssert(replacement_node->outputs().empty(), "Node can't have outputs");
+  DebugAssert(!replacement_node->left_input() && !replacement_node->right_input(), "Node can't have inputs");
+
+  const auto outputs = original_node->outputs();
+  const auto input_sides = original_node->get_input_sides();
+
+  /**
+   * Tie the replacement_node with this nodes inputs
+   */
+  replacement_node->set_left_input(original_node->left_input());
+  replacement_node->set_right_input(original_node->right_input());
+
+  /**
+   * Tie the replacement_node with this nodes outputs. This will effectively perform clear_outputs() on this node.
+   */
+  for (size_t output_idx = 0; output_idx < outputs.size(); ++output_idx) {
+    outputs[output_idx]->set_input(input_sides[output_idx], replacement_node);
+  }
+
+  /**
+   * Untie this node from the LQP
+   */
+  original_node->set_left_input(nullptr);
+  original_node->set_right_input(nullptr);
+}
+
+void lqp_remove_node(const std::shared_ptr<AbstractLQPNode>& node) {
```

Comment:
Not a big fan of these `utils` files that don't have a clear defined purpose. `remove()` for example is something that would easily fit into the node.